tests/cpp/transforms/make_block_local_test.cpp

```diff
@@ -0,0 +1,196 @@
+#include <memory>
+
+#include "gtest/gtest.h"
+#include "taichi/analysis/arithmetic_interpretor.h"
+#include "taichi/ir/ir_builder.h"
+#include "taichi/ir/snode.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/struct/struct.h"
+#include "tests/cpp/struct/fake_struct_compiler.h"
+
+namespace taichi {
+namespace lang {
+namespace {
+
+class MakeBlockLocalTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    root_snode_ = std::make_unique<SNode>(/*depth=*/0, /*t=*/SNodeType::root);
+  }
+
+  void initialize(int pointer_size, int block_size) {
+    // The SNode tree of this test looks like the following
+    //
+    // |root_snode_|  // ti.root
+    // \
+    //  |
+    //  +- |pointer_snode_|  // .pointer(ti.ij, |pointer_size|)
+    //     \
+    //      |
+    //      +- |bls_snode_|  // .dense(ti.ij, |block_size|s)
+    //      |
+    //      +- |struct_for_snode_|  // .dynamic(ti.k, ...)
+    //
+    // The |struct_for_snode_| is the one being iterated over in the struct-for
+    // offloaded task |for_stmt_|, meaning that the loop index is based on this
+    // SNode. On the other hand, |bls_snode_| is the one we want to cache into
+    // the BLS buffer (shared memory).
+    //
+    //
+    // Because |bls_snode_| has a larger shape than |struct_for_snode_|, we want
+    // to see if the tests can handle the loop index scaling multiplier
+    // (block_size) and infer the BLS size correctly.
```

Comment:
nit:
```suggestion
    // |bls_snode_| has a larger shape than |struct_for_snode_|, because we
    // want to see if the tests can handle the loop index scaling multiplier
    // (block_size) and infer the BLS size correctly.
```